### ISSUE TITLE
Have the possibility to rename datasets, tables and columns names

### ISSUE THREAD

**Issue by GoogleCodeExporter** [runt18/google-bigquery#52]
```
BigQuery would be a lot more flexible if it would allow to rename datasets, 
tables and columns.

There are some workarounds but it would be a lot better if we add the option to 
do it.
```

Original issue reported on code.google.com by `collette...@gmail.com` on 18 Nov 2013 at 9:17

**Comment by GoogleCodeExporter** [comment]
```
Any ideas if there is a chance that this will be possible eventually?
```

Original comment by `collette...@gmail.com` on 22 Nov 2013 at 7:18

**Comment by GoogleCodeExporter** [comment]
```
We don't currently have any plans to add these features. Can you tell me more 
about your scenario? Are you concerned about cost, performance, or API 
convenience?

Note that tables and datasets can be renamed by copying them and then deleting 
the originals immediately. Obviously this is not quite as convenient as a 
native rename operation, but at least there's a way to do it.

Columns cannot be renamed quite so easily. Currently the best solution is to 
export to JSON and then re-import, which is slow but free. You could also do 
this by running a query that aliases the column to something else, but you'll 
be charged for scanning the entire table, and if you have nested or repeated 
fields in the table, they'll be flattened in the process.
```

Original comment by `jcon...@google.com` on 22 Nov 2013 at 9:21

**Comment by GoogleCodeExporter** [comment]
```
such feature is very usable for: 
1. maintenance.
2. creating flip-flop mechanism (loading data into temp table and once it is 
ready renaming it)

the suggested solutions are more costly or requires much more effort than an 
atomic meta-data rename operation.

```

Original comment by `Mominisg...@gmail.com` on 11 Dec 2013 at 11:51

**Comment by GoogleCodeExporter** [comment]
```
Thanks for the update. I'll keep this bug open as a feature request, but as 
noted previously, we don't yet have plans to implement this feature.

Note that for the second scenario, a table copy works pretty well. Table copies 
are free (assuming you delete the source table immediately afterward) and fast 
(metadata only in most cases).
```

Original comment by `jcon...@google.com` on 11 Dec 2013 at 4:15

**Comment by GoogleCodeExporter** [comment]
```
Sorry for the delay. 
I would like this feature to be implemented for the same reasons that Mominisg 
mentioned. Also because sometimes we realize that the name that we gave no 
longer makes sense.

Thanks a lot for the reply!
```

Original comment by `collette...@gmail.com` on 11 Dec 2013 at 4:21

**Comment by GoogleCodeExporter** [comment]
```
Cleaning up old issues; the suggestions Jeremy gave in #2 and #4 would be your 
best option for handling these scenarios.
```

Original comment by `thomasp...@google.com` on 22 Aug 2014 at 5:58
- Changed state: **Done**

**Comment by GoogleCodeExporter** [comment]
```
as of #2, I don't think "the best solution is to export to JSON and then 
re-import" is an option. renamed column will just be reported as "no such 
field". The only option is just query and aliases the column.  badly needed 
this feature, lol
```

Original comment by `fionaz...@geotab.com` on 24 Sep 2015 at 8:07

**Comment by GoogleCodeExporter** [comment]
```
I understand the solutions provided in #2 and #4 worked initially, but now that 
BigQuery supports streaming, things become much harder without flip flop to 
perform.

Anybody came out with an idea? My typical use case is the following:

I have a table with a certain datastructure. I want to add a column and 
"update" the existing rows. Of course, I cannot update the rows, so I'm 
planning on creating a new table with the same structure plus the new column, 
then "switch" to have the streaming pushing out to the new table so that I can 
later grab the data from the older table and push it back into the new one. 
Yet, I cannot see a way to safely do this - without changing my system 
configuration or some code - because of the incoming flow of stream.

In other words, I want to perform something similar to this:

RENAME TABLE foo TO foo_old, foo_new TO foo;

For now, I'm thinking of adding a switch in my codebase to forward to another 
table and then switch it back later. Any body have other thoughts?
```

Original comment by `alex...@4-iq.com` on 23 Oct 2015 at 8:38

**Comment by GoogleCodeExporter** [comment]
```
Hello, is there still a plan to add that feature?
Even, just moving a table from a dataset to another would be helpful as long as 
it returns fairly quickly.
Having this type of feature would be very helpful since it would reduce 
down-time when "updating" (by deleting the old table and then loading a new one 
since there is no "update" feature)
Thoughts?
```

Original comment by `moun...@wepay.com` on 24 Nov 2015 at 12:29

**Comment by GoogleCodeExporter** [comment]
```
Re #7: Not sure what you mean. Can you provide more detail?

Re #8: I don't think rename would solve your problem, because it would be 
challenging for a rename implementation to meet the exact atomicity 
requirements you've specified. I think the best solution is to point your 
streaming ingestion to a new table while you modify/update the old one. You may 
even consider streaming to a new table each day.

Re #9: You can copy from one table to another with writeDisposition set to 
WRITE_TRUNCATE, which will atomically overwrite your destination table with the 
source table's contents. That's the best way to get an atomic update of a table.
```

Original comment by `jcon...@google.com` on 7 Dec 2015 at 5:52

**Comment by GoogleCodeExporter** [comment]
```
Issue 383 has been merged into this issue.
```

Original comment by `jcon...@google.com` on 7 Dec 2015 at 5:53
